### PR TITLE
#18 Bugfix for empty data

### LINKED ISSUE #18: Include the 0 partition fix
## Describe the bug
The bug was found in Enceladus [here](https://github.com/AbsaOSS/enceladus/pull/1991)

## Expected behavior
Include those changes

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #18

### CODE REVIEW

File: src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala [hunk 1]
```diff
@@ -56,26 +56,39 @@ object DataFramePartitioner {
...
-        case (None, Some(max)) if currentBlockSize > max =>
-          df.repartition(computeBlockCount(sizeInBytes, max, addRemainder = true))
-        case (Some(min), Some(max)) if currentBlockSize < min || currentBlockSize > max =>
-          df.repartition(computeBlockCount(sizeInBytes, max, addRemainder = true))
-        case _ => df
+        val currentBlockSize = sizeInBytes / df.rdd.getNumPartitions
+
+        (minPartitionSize, maxPartitionSize) match {
+          case (Some(min), None) if currentBlockSize < min =>
+            changePartitionCount(computeBlockCount(min, sizeInBytes, addRemainder = false), df.coalesce)
```
Comment (dk1844) [inline src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala:82]: Using `coalesce` here is smart 👍

File: src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala [hunk 2]
```diff
@@ -56,26 +56,39 @@ object DataFramePartitioner {
 
     def repartitionByPlanSize(minPartitionSize: Option[ByteSize], maxPartitionSize: Option[ByteSize]): DataFrame = {
 
-      def computeBlockCount(totalByteSize: BigInt, desiredSize: Long, addRemainder: Boolean): Int = {
+      def computeBlockCount(totalByteSize: BigInt, desiredSize: BigInt, addRemainder: Boolean): Int = {
```
Comment (dk1844) [inline src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala:59]: What is the reason for this change? (just wondering, it does not seem to make much of a difference, but maybe I am missing something)

**Review by dk1844 — DISMISSED** [review]
LGTM (just read the code)

File: src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala (hunk 2 above)
Comment (AdrianOlosutean) [reply src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala:59]: `sizeInBytes` is a BigInt and although in Enceladus it worked when passing a BigInt as a Long param, here it seems to fail compilation

File: src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala (hunk 1 above)
Comment (AdrianOlosutean) [reply src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala:82]: Still probably debateable 😅

**Review by dk1844 — APPROVED** [review]
LGTM, read the code again.

File: src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala (hunk 1 above)
Comment (benedeki) [reply src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala:82]: I actually think it's not desired. (I deliberately put `repartition` here.)
AFAIK `coalesce` just "put's together" some partitions to reach the desired number. I think the gola of the function in general should be to create roughly uniform partitions (unless there will be some special paritionings in the future). And with `coalesce` it's not gonna happen unless some further analysis is done (might be worth for those cases).

Comment (dk1844) [reply src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala:82]: Definitely debatable. Well, `coalesce` can result in _roughly_ uniform partitions, it's just that this _roughly_ may or may not be good enough. `repartition` will yield similar or better uniformity in exchange for more work. 🤷

**Pull request merged**